File: apps/station/src/lib/close-pallet.ts
import { buildSscc, DomainError } from "@markiro/domain";
import { closePallet, currentPallet } from "./pallets.js";
import { burnSerial } from "./sscc-pool.js";
import type { CloseBoxDeps } from "./close-box.js";
import type { CredentialGeneration } from "./credential-recovery.js";
import { acquireCredentialCommitLease } from "./credential-recovery.js";
import {
  StationGrantAdmission,
  stationOperatorIsCurrentlyActive,
} from "./offline-grants/admission.js";
import { sampleGrantClock, type GrantClockSample } from "./offline-grants/clock.js";
import { readShiftExecutionProjection } from "./offline-grants/semantic.js";
import { OfflineGrantDeniedError } from "./journal.js";

/**
 * Extension digit for pallet serial ranges. Boxes take 0 -- see
 * `sscc-pool.ts` for why one shared space would put the same number on a
 * box and on the pallet it stands on.
 */
export const PALLET_EXTENSION_DIGIT = 1;

export type ClosePalletResult =
  | { status: "closed"; palletId: string; sscc: string; boxCount: number; closedAt: string }
  | { status: "no-serials" }
  | { status: "empty" }
  /**
   * Mirrors `close-box.ts`'s own `invalid-serial`: the burned serial cannot
   * build a valid SSCC (`buildSscc` threw `SSCC_RANGE`). The serial is
   * already burned -- `burnSerial` is one atomic statement, by design, so
   * there is no way to give it back -- and is accepted as lost, the same
   * way an abandoned pallet already costs a burned serial. The pallet row
   * is left untouched (still open, no sscc/closedAt written) so the
   * operator can simply try closing it again.
   */
  | { status: "invalid-serial" }
  /**
   * `closePallet`'s `closed_at IS NULL` guard matched nothing: another call
   * already closed THIS pallet, with a different SSCC, between the
   * `currentPallet` read above and the UPDATE -- a genuine double close, by
   * two concurrent closers racing for the same pallet. Unlike
   * `invalid-serial`, the row was NOT left untouched: it already carries a
   * different, already-persisted SSCC written by the winning call. A caller
   * must not retry closing "the current pallet" here -- by the time it
   * reads again, this pallet is no longer open, so a naive retry would find
   * a different, not-yet-full pallet (or none) and silently burn another
   * serial against it. The serial burned by THIS losing call cannot be
   * un-burned and never reached a stored record, so it is accepted as lost,
   * the same trade `invalid-serial` already names above.
   */
  | { status: "already-closed" };

/**
 * Closes the shift's current open pallet.
 *
 * The order of the checks is the design, exactly as in `close-box.ts`:
 * emptiness is tested BEFORE burning, so a pallet closed by mistake costs no
 * serial, and the serial is burned only once the pool actually yields one.
 * A pallet abandoned at shift end therefore costs nothing either.
 */
async function closeCurrentPalletLegacy(
  deps: CloseBoxDeps,
  shiftId: string,
  operatorId: string | null,
): Promise<ClosePalletResult> {
  const pallet = await currentPallet(deps.exec, shiftId, deps.terminalId);
  if (!pallet || pallet.boxCount === 0) return { status: "empty" };

  const serial = await burnSerial(deps.exec, deps.issuerPrefix, PALLET_EXTENSION_DIGIT);
  if (serial === null) return { status: "no-serials" };

  let sscc: string;
  try {
    sscc = buildSscc(PALLET_EXTENSION_DIGIT, deps.issuerPrefix, serial);
  } catch (err) {
    if (err instanceof DomainError && err.code === "SSCC_RANGE") {
      // The serial is already burned and cannot be given back -- `burnSerial`
      // is one atomic statement by design -- and is accepted as lost, exactly
      // as an abandoned pallet already costs one. The pallet row is left
      // untouched so the operator can try again.
      return { status: "invalid-serial" };
    }
    throw err;
  }

  const closedAt = new Date(deps.now ? deps.now() : Date.now()).toISOString();
  const closed = await closePallet(deps.exec, pallet.palletId, sscc, closedAt, operatorId);
  if (!closed) {
    // `closePallet`'s `closed_at IS NULL` guard matched nothing: another
    // call already closed this exact pallet between the `currentPallet`
    // read above and this UPDATE. `closePallet` is one guarded UPDATE, not
    // a multi-statement sequence, so this is NOT the pooled-connection race
    // `apps/station/AGENTS.md` warns about (that hazard is about separate
    // calls landing on different connections inside what should be one
    // transaction). The real cause here is a genuine double close: two
    // concurrent callers both read the same open pallet and both tried to
    // close it. The serial burned above cannot be un-burned and never
    // reached a stored record, so it is accepted as lost -- exactly the
    // trade `invalid-serial` already names above -- but this is reported as
    // its own `already-closed` status rather than `invalid-serial`: the row
    // was NOT left untouched here the way `invalid-serial` promises: it was
    // closed by the winning caller with a different, already-persisted
    // SSCC, so a caller must not retry closing "the current pallet".
    return { status: "already-closed" };
  }
  return { status: "closed", palletId: pallet.palletId, sscc, boxCount: pallet.boxCount, closedAt };
}

/** Legacy entry point cannot bypass an installed grant owner. */
export async function closeCurrentPallet(
  deps: CloseBoxDeps,
  shiftId: string,
  operatorId: string | null,
): Promise<ClosePalletResult> {
  const [active] = await deps.exec.all<{ active: number }>(
    "SELECT 1 active FROM offline_grant_install_state WHERE id=1",
  );
  if (active?.active === 1) throw new OfflineGrantDeniedError("grant_aware_owner_required");
  return closeCurrentPalletLegacy(deps, shiftId, operatorId);
}

/** Grant-aware pallet owner: allowance, serial CAS and pending-print close are one command. */
export async function closeCurrentPalletWithOfflineGrant(
  deps: CloseBoxDeps,
  shiftId: string,
  operatorId: string | null,
  generation: CredentialGeneration,
  clock: () => Promise<GrantClockSample> = sampleGrantClock,
): Promise<ClosePalletResult> {
  const lease = acquireCredentialCommitLease(generation);
  if (!lease) throw new OfflineGrantDeniedError("stale_credential");
  try {
    const [state] = await deps.exec.all<{
      tenant_id: string;
      device_id: string;
      owner_kind: "station";
      credential_epoch: number;
    }>(
      "SELECT tenant_id,device_id,owner_kind,credential_epoch FROM offline_grant_install_state WHERE id=1",
    );
    if (!state) return closeCurrentPalletLegacy(deps, shiftId, operatorId);
    if (!operatorId) throw new OfflineGrantDeniedError("operator_unauthorized");
    if (!(await stationOperatorIsCurrentlyActive(deps.exec, operatorId)))
      throw new OfflineGrantDeniedError("operator_unauthorized");
    const pallet = await currentPallet(deps.exec, shiftId, deps.terminalId);
    if (!pallet || pallet.boxCount === 0) return { status: "empty" };
    const [pool] = await deps.exec.all<{ rowid: number; serial: number }>(
      "SELECT rowid,next_serial serial FROM sscc_pool WHERE issuer_prefix=? AND extension_digit=? AND next_serial<=to_serial ORDER BY from_serial LIMIT 1",
      [deps.issuerPrefix, PALLET_EXTENSION_DIGIT],
    );
    if (!pool) return { status: "no-serials" };
    let sscc: string;
    try {
      sscc = buildSscc(PALLET_EXTENSION_DIGIT, deps.issuerPrefix, pool.serial);
    } catch (error) {
      if (error instanceof DomainError && error.code === "SSCC_RANGE")
        return { status: "invalid-serial" };
      throw error;
    }
    const [binding] = await deps.exec.all<{ snapshot_digest: string }>(
      "SELECT json_extract(grant_json,'$.snapshotDigest') snapshot_digest FROM offline_grant_grants WHERE json_extract(grant_json,'$.kindOfGrant')='task' AND json_extract(grant_json,'$.taskKind')='shift' AND json_extract(grant_json,'$.taskId')=? ORDER BY installed_sequence DESC LIMIT 1",
      [shiftId],
    );
    const eventId = crypto.randomUUID();
    const closedAt = new Date(deps.now ? deps.now() : Date.now()).toISOString();
    const result: ClosePalletResult = {
      status: "closed",
      palletId: pallet.palletId,
      sscc,
      boxCount: pallet.boxCount,
      closedAt,
    };
    const committed = await new StationGrantAdmission(deps.exec, clock).commitCompletion({
      operatorId,
      intent: {
        owner: {
          tenantId: state.tenant_id,
          deviceId: state.device_id,
          kind: state.owner_kind,
          credentialEpoch: state.credential_epoch,
        },
        capability: "shift.start.v1",
        taskId: shiftId,
        snapshotDigest: binding?.snapshot_digest ?? "missing",
        eventId,
        eventType: "shift.pallet.close.v1",
        cost: {},
      },
      execution: await readShiftExecutionProjection(deps.exec, shiftId),
      event: {
        eventId,
        shiftId,
        palletId: pallet.palletId,
        sscc,
        boxCount: pallet.boxCount,
        closedAt,
        operatorId,
        terminalId: deps.terminalId,
      },
      facts: { containers: 1 },
      result,
      wrapCommand(command) {
        return {
          sql: "INSERT INTO offline_grant_pallet_close_commands(event_id,payload_json) VALUES(?,?)",
          values: [
            eventId,
            JSON.stringify({
              grantCommand: JSON.parse(String(command.values[1])) as unknown,
              poolRowId: pool.rowid,
              serial: pool.serial,
              palletId: pallet.palletId,
              sscc,
              closedAt,
              operatorId,
            }),
          ],
        };
      },
    });
    if (!committed.decision.allow)
      throw new OfflineGrantDeniedError(committed.decision.reason ?? "denied");
    return committed.result as ClosePalletResult;
  } finally {
    lease.release();
  }
}

File: apps/handheld/app/src/main/kotlin/app/markiro/handheld/core/box/ClosePallet.kt
package app.markiro.handheld.core.box

import androidx.room.withTransaction
import app.markiro.handheld.core.storage.HandheldDatabase
import app.markiro.handheld.core.storage.PalletEntity
import app.markiro.handheld.core.storage.PalletPrint
import app.markiro.handheld.core.util.Iso

sealed interface ClosePalletResult {
    data class Closed(
        val pallet: PalletEntity,
        val sscc: String,
        val boxCount: Int,
        val closedAt: String,
    ) : ClosePalletResult

    /** No pallet is open, or the open one holds no boxes. Nothing was burned. */
    data object Empty : ClosePalletResult

    /** This issuer prefix's pool is dry. Nothing was burned and the pallet stays open. */
    data object NoSerials : ClosePalletResult

    /**
     * The burned serial cannot become a valid SSCC.
     *
     * Only reachable when this device holds a range reaching past its issuer
     * prefix's own capacity — a corrupted local pool. The serial is already
     * burned and cannot be given back, exactly as an abandoned pallet costs
     * one, and the pallet is left OPEN so the operator can try again rather
     * than being stranded half-closed. Surfaced by name so it is not repeated
     * silently until the whole invalid block is spent.
     */
    data object InvalidSerial : ClosePalletResult

    /** The shift carries no SSCC issuer, so no pallet of it can ever be numbered. */
    data object NoIssuer : ClosePalletResult
}

/**
 * Closes the shift's open pallet.
 *
 * `CloseBox` one level up, keeping every one of its invariants for the exact
 * same reasons: emptiness is checked BEFORE burning, so a pallet closed
 * before any box joined it (or closed by mistake) costs no serial; burning
 * and closing are ONE `db.withTransaction`, because a serial that leaves the
 * pool without landing on a pallet is gone and the pool has no way to give
 * one back; and the `AlreadyClosed` rollback takes the burn back with it if
 * the guarded close affects no row.
 *
 * `CloseBox` calls this automatically once a box closing brings the open
 * pallet to capacity, and the operator's own «Закрыть паллету досрочно» calls
 * it directly. Those are separate coroutines, so without the shared
 * [PalletLock] both could pass the open-pallet read together and each burn a
 * serial for the same pallet -- identical to why `CloseBox` guards itself
 * against its own automatic-close-at-capacity race.
 *
 * Both entry points take that lock BEFORE opening the transaction below, and
 * neither runs inside somebody else's: `CloseBox` commits its own box
 * transaction first and calls [close] with the lock still in hand. That single
 * order is what keeps the automatic and manual paths out of an ABBA deadlock,
 * and it also keeps the `AlreadyClosed` rollback local -- a nested
 * `withTransaction` that throws marks the OUTER transaction for rollback on
 * Android's SQLiteDatabase even when the exception is caught here.
 */
class ClosePallet(
    private val db: HandheldDatabase,
    private val pool: SsccPool,
    private val lock: PalletLock,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    /** Rolls the burn back when the guarded close turns out to affect no row. */
    private class AlreadyClosed : Exception()

    /** Closes the shift's open pallet, taking the pallet lock itself. */
    suspend fun close(shiftId: String, issuerPrefix: String?, operatorId: String?): ClosePalletResult =
        lock.withLock { held -> close(held, shiftId, issuerPrefix, operatorId) }

    /**
     * [close] for a caller that already holds the lock -- `CloseBox`, which
     * holds it across the box transaction it just committed so no other
     * coroutine can close this pallet, or open a different one, in between.
     */
    suspend fun close(
        held: PalletLock.Held,
        shiftId: String,
        issuerPrefix: String?,
        operatorId: String?,
    ): ClosePalletResult {
        lock.requireHeld(held)
        lock.requireNoTransaction("Closing a pallet")
        if (issuerPrefix == null) return ClosePalletResult.NoIssuer
        val pallet = db.palletDao().open(shiftId) ?: return ClosePalletResult.Empty
        val boxCount = db.palletDao().boxCount(pallet.palletId)
        if (boxCount == 0) return ClosePalletResult.Empty

        // Burning and closing are ONE transaction. A serial that leaves the
        // pool without landing on a pallet is gone -- the pool has no way to
        // give one back -- so the guarded update failing has to take the
        // burn with it.
        return try {
            db.withTransaction {
                val serial = pool.burn(issuerPrefix, SsccPool.PALLET_EXTENSION_DIGIT)
                    ?: return@withTransaction ClosePalletResult.NoSerials
                val sscc = try {
                    Sscc.build(SsccPool.PALLET_EXTENSION_DIGIT, issuerPrefix, serial)
                } catch (_: SsccException) {
                    // The serial IS spent here, deliberately: see CloseBox's
                    // identical InvalidSerial case for why rolling back
                    // would only hand the same impossible serial out again.
                    return@withTransaction ClosePalletResult.InvalidSerial
                }
                // The pallet's own moment, persisted: the label's «Дата
                // производства» and «Годен до» derive from it, and a
                // recovery print the next morning must stamp the same two
                // dates rather than that morning's.
                val closedAt = Iso.format(clock())
                if (db.palletDao().close(pallet.palletId, sscc, closedAt, operatorId) == 0) {
                    throw AlreadyClosed()
                }
                ClosePalletResult.Closed(
                    pallet = pallet.copy(
                        sscc = sscc,
                        closedAt = closedAt,
                        operatorId = operatorId,
                        printState = PalletPrint.PENDING,
                        printReason = null,
                    ),
                    sscc = sscc,
                    boxCount = boxCount,
                    closedAt = closedAt,
                )
            }
        } catch (_: AlreadyClosed) {
            ClosePalletResult.Empty
        }
    }
}
